[Wallet] Ensure <0.13 clients can't open HD wallets

jonasschnelli opened the pull request:
In current master, you could run a 0.13 HD wallet in a non HD compatible 0.12 client, resulting in a mix of HD and non HD key. You could even refill the wallets keypool running a 0.12 client and use it on 0.13 which would lead to a `getnewaddress` call in 0.13 responding a non HD address.

This PR ensures that HD wallets will at least require version 0.12.99.

This only affects newly created wallets and only if the user had not passed `-usehd=0` (old <0.13 created wallets will still run on <0.13 clients once they where opened in 0.13).

jonasschnelli committed:
[Wallet] Ensure <0.13 clients can't open HD wallets

MarcoFalke commented:
Concept ACK

MarcoFalke commented:
quickly tested ACK 51a4ee8

MarcoFalke commented on src/wallet/wallet.h:
```diff
@@ -78,7 +78,8 @@ enum WalletFeature
     FEATURE_COMPRPUBKEY = 60000, // compressed public keys
 
-    FEATURE_LATEST = 60000
+    FEATURE_HD = 129900, // Hierarchical key derivation after BIP32 (HD Wallet)
```
Strictly speaking not all 12.99's support HD, so we may want to adjust to 130000 after the branch off.

laanwj replied:
This isn't derived from the actual version, is it?

MarcoFalke replied:
It is the client version: https://dev.visucore.com/bitcoin/doxygen/class_c_wallet.html#a3e37adcf1b2ad9fc02ae3e8da7c692c9

jonasschnelli replied:
Yes. We can try to merge this once we split of the masterbranch and had increase the version.

laanwj replied:
Yes using the client version here is not very nice, if we had a dedicated wallet version constant this problem could have been avoided, as it could have been bumped at the moment HD support was introduced instead of arbitrary delineation based on release-process constraints.

But for now that's the best solution.

MarcoFalke replied:
@jonasschnelli Could you already bump this so we don't lose time later? (In case you are not around when the branch off happens)

jonasschnelli replied:
@MarcoFalke 
If I bump this to 130000 now, HD would no longer work on current master IMO.

For <0.13 incompatibility, we need to stick to the current check based on `CLIENT_VERSION`.
https://github.com/bitcoin/bitcoin/blob/master/src/wallet/walletdb.cpp#L637

Once we change the wallet database format, we could introduce a set of strings about the features the wallet database requires/supports.

MarcoFalke replied:
> If I bump this to 130000 now, HD would no longer work on current master IMO.

True, but I think the plan was to merge this after the bump to 13.99 into master?

MarcoFalke replied:
No strong opinion, though.

laanwj replied:
Version have been bumped, you can update this now

laanwj committed:
build: bump version to 0.13.0

laanwj committed:
build: Release notes update

Fill in the header, and move items to the appropriate part of the
release notes structure.

jtimon commented:
ut ACK 51a4ee8

MarcoFalke commented on doc/release-notes.md:
```diff
@@ -197,6 +197,8 @@ You can't disable HD key generation once you have created a HD wallet.
 
 There is no distinction between internal (change) and external keys.
 
+HD wallets are incompatible with older versions of Bitcoin Core.
```
I think you need to open a separate pull against .13 now that the release notes were moved.

jonasschnelli commented:
Sorry. Have to close and reopen (new PR) this because use this local branch for #8366.

jtimon commented:
Was a replacement PR opened?
